### PR TITLE
[None][perf] Pre-JIT Mamba SSD HAS_INITSTATES=True kernels during warmup

### PR DESCRIPTION
**Pull request by chenfeiz0326**
## Summary
- Pre-JIT the `mamba_chunk_scan_combined` + `_state_passing_fwd` Triton kernels with `HAS_INITSTATES=True` during `PyTorchModelEngine._warmup`, so the first prefill that carries a cached prefix does not pay the compile cost inline.
- Fixes a ~20% mid-run latency spike (bench iter ~2000) on Nemotron-Nano-12B-v2 that presents as 3-rep total_token_throughput instability (CV ~14%).

The warmup hook is opt-in per module (`hasattr(module, "warmup_ssd_initstates_kernels")`) and wrapped in a try/except that only logs — non-Mamba models are unaffected.

## Verification
On bia B300, `nemotron_nano_12b_v2-bench-pytorch-streaming-bfloat16-maxbs:512-maxnt:2048-input_output_len:500,2000-con:250`:

|            | rep1 | rep2 | rep3 | median | CV     |
|------------|------|------|------|--------|--------|
| Before     | 7481 | 9472 | 9612 | 9472   | 13.9%  |
| After (this PR) | 9580 | 9788 | 9794 | 9788   | 1.25%  |

- Median gain: +3.36% over pre-fix rep3; +3.59% over cached ToT median (9449.36 t/s).
- Nsys trace on the fix wheel shows `_state_passing_fwd_kernel` at 0.153 ms across all 28 prefill invocations (no first-call compile spike).

## Test plan
- [x] CI: relevant Mamba2 unit tests still pass.
- [x] Bench Nemotron-Nano-12B-v2 on B200/B300 3-rep and confirm CV ≤ 5%.
- [x] Confirm no regression on non-Mamba models (warmup hook is guarded by `hasattr` on each `nn.Module`).

## Summary by CodeRabbit

* **New Features**
  * Added an additional warmup step for Mamba-based models to prepare specialized kernels ahead of first use.
  * Warmup now covers more execution paths, helping improve startup readiness for supported Mamba workloads.

* **Bug Fixes**
  * Improved warmup robustness by safely skipping the new step if the environment does not support it, while logging a warning instead of failing.

### CODE REVIEW

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "DGX_B200-8_GPUs-PyTorch-PerfSanity-Post-Merge-6,GB300-36_GPUs-9_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU2-GEN1-NODE8-GPU32-Post-Merge-1,DGX_B200-8_GPUs-PyTorch-PerfSanity-Post-Merge-3"

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "DGX_B200-8_GPUs-PyTorch-PerfSanity-Post-Merge-6,GB300-36_GPUs-9_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU2-GEN1-NODE8-GPU32-Post-Merge-1,DGX_B200-8_GPUs-PyTorch-PerfSanity-Post-Merge-3"

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py [hunk 1]
```diff
@@ -282,6 +283,81 @@ def _try_attach_nvfp4_scale(self):
...
+                dt_limit=(0.0, float("inf")),
+                cu_seqlens=cu_seqlens,
+                seq_idx=seq_idx,
+                return_varlen_states=True,
+                return_final_states=False,
+                out=out_buf,
+                state_dtype=state_dtype,
+            )
+            torch.cuda.synchronize()
+        except Exception as e:
```
Comment (achartier) [inline tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: +1 to Coderabbit, something like a CUDA OOM should be fatal.

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py [hunk 2]
```diff
@@ -282,6 +283,81 @@ def _try_attach_nvfp4_scale(self):
...
+        # step that carries a non-empty cached prefix pays a one-shot
+        # kernel-compile cost that shows up as a ~20% latency spike in
+        # bench iteration ~2000 (Nemotron-Nano-12B-v2, bia B300).
+        try:
+            weight = self.conv1d.weight
+            device = weight.device
+            in_dtype = weight.dtype
+            state_dtype = self._mamba_ssm_cache_dtype or in_dtype
+            num_prefills = 2
+            seq_per = int(self.chunk_size) * 2
```
Comment (achartier) [inline tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: `chunk_size` is already a int

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py [hunk 3]
```diff
@@ -282,6 +283,81 @@ def _try_attach_nvfp4_scale(self):
...
+                seq_per).unsqueeze(0)
+            chunk_indices, chunk_offsets = (
+                cu_seqlens_to_chunk_indices_offsets_triton(
+                    cu_seqlens=cu_seqlens,
+                    chunk_size=int(self.chunk_size),
+                    total_seqlens=total))
+            out_buf = torch.empty((1, total, self.tp_nheads, self.head_dim),
+                                  dtype=in_dtype,
+                                  device=device)
+            mamba_chunk_scan_combined(
```
Comment (achartier) [inline tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: Ideally, we should refactor to avoid code duplication with `forward`

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py [hunk 4]
```diff
@@ -282,6 +283,81 @@ def _try_attach_nvfp4_scale(self):
...
+                dt_softplus=self.delta_softplus,
+                dt_limit=(0.0, float("inf")),
+                cu_seqlens=cu_seqlens,
+                seq_idx=seq_idx,
+                return_varlen_states=True,
+                return_final_states=False,
+                out=out_buf,
+                state_dtype=state_dtype,
+            )
+            torch.cuda.synchronize()
```
Comment (achartier) [inline tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: Could be a simple `current_stream().synchronize()`

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 5]
```diff
@@ -1090,6 +1090,17 @@ def warmup(self, resource_manager: ResourceManager) -> None:
             self._general_warmup(resource_manager, warmup_requests_configs)
             log_mem_snapshot("warmup/after_memory_pool_prepop")
 
+        # Pre-JIT Mamba SSD HAS_INITSTATES=True Triton kernels so the first
+        # prefill that carries a cached prefix does not pay the compile cost
+        # inline. Runs for all cache-manager kinds (incl. MambaHybridCacheManager),
+        # since that manager is exactly the one that exposes the cold path.
+        model_root = getattr(self, "model", None)
```
Comment (achartier) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: `model` being None here should be an error?

File: tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py (hunk 3 above)
Comment (chenfeiz0326) [reply tensorrt_llm/_torch/modules/mamba/mamba2_mixer.py]: Extracted a small `_run_ssd_scan` helper in b8da7f8 that the warmup path now calls, so the kernel-call kwargs live in one place. I intentionally left `forward()` calling `mamba_chunk_scan_combined` directly for this PR — its call site uses a view into the preallocated output buffer (`preallocated_ssm_out_p`) rather than a fresh `out_buf`, and `state_dtype` is derived from a different path. Wiring `forward()` through the helper is a real hot-path refactor that I'd rather do as a follow-up so this warmup-only PR stays narrow. Happy to file the follow-up if you'd prefer.

**Review by achartier — APPROVED** [review]

**Comment by chenfeiz0326** [comment]
/bot run --disable-fail-fast --stage-list "DGX_B200-8_GPUs-PyTorch-PerfSanity-Post-Merge-3,GB300-36_GPUs-9_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU2-GEN1-NODE8-GPU32-Post-Merge-1,GB300-8_GPUs-2_Nodes-PyTorch-Disagg-PerfSanity-CTX1-NODE1-GPU4-GEN1-NODE1-GPU4-Post-Merge-3"

**Comment by chenfeiz0326** [comment]
Superseded by #16177, which fixes both this case (nemotron_nano_12b_v2 on B300) and the nemotron_3_super_120b run-1 warmup deficit across GB200/GB300/B300 with a single autotuner-warmup path. Closing this PR — see #16177 for the 5-case verification table.
